mailbox/api/src/main/java/org/apache/james/mailbox/model/UidValidity.java

```diff
@@ -0,0 +1,61 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.model;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
+
+public class UidValidity {
```

Comment:
you are missing the actual definition of UIDVALIDITY from IMAP RFC : 
```
nz-number       = digit-nz *DIGIT
                    ; Non-zero unsigned 32-bit integer
                    ; (0 < n < 4,294,967,296)
```
Sadly we don't have unsigned 32-bit integer in Java, it's why we decided to use a long (that can contain the range [0-2^32]). However, you should ensure we can't create a UidValidity object that is outside that range.

Reply:
Not that easy:

```

    protected int randomUidValidity() {
        return Math.abs(RANDOM.nextInt());
    }
```

We created many unvalid UID VALIDITY, we need to be still lenient and accept already created fields. However newer UID VALIDITY should match condition matthieu highlight.

Reply:
You assume returning a negative integer worked. I would argue it did not and we never noticed.

1 minute search : https://github.com/GNOME/evolution/commit/e81e64f8dd9facde47e93e3f4e8b190978eeb91b#diff-e4a32bc822c57a9a503008fe6358736fR100

Let's fix it. It happens that we can change the existing values if we detect they were invalid and it will trigger a client re-sync. Not that bad, right?

Reply:
Not that bad, but this needs low level support (update if invalid) on every backend.

This is expensive to implement.

So I would:

 - Create a dedicated JIRA ticket for this "invalid UID_VALIDITY are generated"
 - Enact the **tactical move**: merge the strong typing + ensure we don't generate new invalid values. This will at list help with new deployments.
 - Document the **strategical move**: fix all and every invalid UID_VALIDITY (at the cost of resynchronicsation)

Would you agree @mbaechler ?

Reply:
Of course. A small step now is better than nothing at all.